Test: herddb.server.hammer.DirectMultipleConcurrentUpdatesTest.testWithCheckpointsAndIndexes is failing on CI

**BUG REPORT**

1. Please describe the issue you observed:

- What did you do?
Checked CI results

- What did you expect to see?
All tests passing

- What did you see instead?

One tests failed:
herddb.server.hammer.DirectMultipleConcurrentUpdatesTest.testWithCheckpointsAndIndexes


```
Mar 19, 2019 9:35:27 PM herddb.core.TableManager checkpoint
INFO: checkpoint mytable finished, logpos (1,2671), 34 active pages, 15 dirty pages, flushed 1,195 records, total time 155 ms
Mar 19, 2019 9:35:27 PM herddb.core.TableSpaceManager checkpoint
INFO: checkpoint done for table herd.mytable (pin: false)
Mar 19, 2019 9:35:27 PM herddb.file.FileDataStorageManager writeCheckpointSequenceNumber
INFO: checkpoint for 9da80a25db214b2f8e0c91a732d315f0 at (1,2671) to /data/tmp/junit8749400299071688528/junit4961568361934239704/9da80a25db214b2f8e0c91a732d315f0.tablespace/checkpoint.1.2671.checkpoint
Mar 19, 2019 9:35:27 PM herddb.file.FileCommitLog dropOldLedgers
SEVERE: dropOldLedgers herd lastCheckPointSequenceNumber: (1,2671), currentLedgerId: 1
Mar 19, 2019 9:35:27 PM herddb.file.FileCommitLog dropOldLedgers
SEVERE: Deleted logfiles: 0
Mar 19, 2019 9:35:27 PM herddb.core.TableSpaceManager checkpoint
INFO: gatto checkpoint finish herd started ad (1,2671), finished at (1,2671), total time 531 ms
Mar 19, 2019 9:35:29 PM herddb.core.TableSpaceManager checkpoint
INFO: gatto checkpoint start herd at (1,3938)
Mar 19, 2019 9:35:29 PM herddb.core.TableManager checkpoint
INFO: tableCheckpoint dirtyThreshold: 0.25, 9da80a25db214b2f8e0c91a732d315f0.mytable (pin: false)
Mar 19, 2019 9:35:29 PM herddb.core.DBManager executePlanAsync
SEVERE: uncaught error
herddb.storage.DataStorageManagerException: error reading data page: 9da80a25db214b2f8e0c91a732d315f0_c49e1e8d-8c85-47fa-adfd-a1bf76855742.100
	at herddb.file.FileDataStorageManager.readPage(FileDataStorageManager.java:261)
	at herddb.core.TableManager.temporaryLoadPageToMemory(TableManager.java:1952)
	at herddb.core.TableManager.fetchDataPage(TableManager.java:3299)
	at herddb.core.TableManager.fetchRecord(TableManager.java:3227)
	at herddb.core.TableManager.accessRecord(TableManager.java:3175)
	at herddb.core.TableManager.lambda$streamTableData$23(TableManager.java:3102)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.Streams$StreamBuilderImpl.tryAdvance(Streams.java:397)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:294)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:300)
	at java.base/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
	at herddb.core.StreamDataScanner.fetchNext(StreamDataScanner.java:53)
	at herddb.core.StreamDataScanner.<init>(StreamDataScanner.java:44)
	at herddb.core.TableManager.scanWithStream(TableManager.java:2937)
	at herddb.core.TableManager.scan(TableManager.java:2675)
	at herddb.core.TableSpaceManager.scan(TableSpaceManager.java:553)
	at herddb.core.DBManager.scan(DBManager.java:756)
	at herddb.core.DBManager.executePlanAsync(DBManager.java:687)
	at herddb.core.DBManager.executePlan(DBManager.java:664)
	at herddb.core.TestUtils.scan(TestUtils.java:71)
	at herddb.server.hammer.DirectMultipleConcurrentUpdatesTest.performTest(DirectMultipleConcurrentUpdatesTest.java:215)
	at herddb.server.hammer.DirectMultipleConcurrentUpdatesTest.testWithCheckpointsAndIndexes(DirectMultipleConcurrentUpdatesTest.java:102)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:379)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413)
Caused by: java.io.EOFException: array len 0, pos 0, try to read 1 bytes
	at herddb.utils.ByteArrayCursor.checkReadable(ByteArrayCursor.java:99)
	at herddb.utils.ByteArrayCursor.readByte(ByteArrayCursor.java:93)
	at herddb.utils.ByteArrayCursor.readVLong(ByteArrayCursor.java:171)
	at herddb.utils.ByteArrayCursor.readVLong(ByteArrayCursor.java:156)
	at herddb.file.FileDataStorageManager.rawReadDataPage(FileDataStorageManager.java:278)
	at herddb.file.FileDataStorageManager.readPage(FileDataStorageManager.java:255)
	... 50 more

```

--- a/herddb-core/src/main/java/herddb/core/TableManager.java
+++ b/herddb-core/src/main/java/herddb/core/TableManager.java
@@ -39,7 +39,6 @@ import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.atomic.LongAdder;
@@ -105,6 +104,7 @@ import herddb.storage.DataStorageManagerException;
 import herddb.storage.FullTableScanConsumer;
 import herddb.storage.TableStatus;
 import herddb.utils.BatchOrderedExecutor;
+import herddb.utils.BooleanHolder;
 import herddb.utils.Bytes;
 import herddb.utils.DataAccessor;
 import herddb.utils.EnsureLongIncrementAccumulator;
@@ -686,9 +686,9 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
     }
 
     /**
-     * Returns currently loaded pages id. To be used only for test to inspect current pages!
+     * Returns currently loaded pages. To be used only for test to inspect current pages!
      */
-    Collection<DataPage> getLoadedPages() {
+    public Collection<DataPage> getLoadedPages() {
         return pages.values();
     }
 
@@ -911,21 +911,26 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
                 }
             }
 
+            dataStorageManager.writePage(tableSpaceUUID, table.uuid, page.pageId, page.getRecordsForFlush());
+
+            /* Set the page as a fully active page */
+            pageSet.pageCreated(page.pageId, page);
+
             if (keepPageInMemory) {
                 /* If we must keep the page in memory we "covert" the page to immutable */
                 pages.put(page.pageId, page.toImmutable());
 
-                /* And we load to page replacement polcy */
+                /*
+                 * And we load to page replacement policy. This is a critic point: after adding page to page
+                 * replacement policy knowledge it can be unloaded from another thread, we should finished any work
+                 * on the page before of that.
+                 */
                 final Page.Metadata unload = pageReplacementPolicy.add(page);
                 if (unload != null) {
                     unload.owner.unload(unload.pageId);
                 }
             }
 
-            /* Set the page as a fully active page */
-            pageSet.pageCreated(page.pageId, page);
-
-            dataStorageManager.writePage(tableSpaceUUID, table.uuid, page.pageId, page.getRecordsForFlush());
 
         } finally {
 
@@ -1944,26 +1949,40 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
 
     }
 
+    /**
+     * Just read a page from {@link DataStorageManager} and return it as an immutable page.
+     *
+     * @param pageId id of page to be loaded
+     * @return loaded page
+     * @throws DataStorageManagerException   if requested page cannot be read
+     */
     private DataPage temporaryLoadPageToMemory(Long pageId) throws DataStorageManagerException {
-        long _start = System.currentTimeMillis();
-        List<Record> page;
+
+        long start = System.currentTimeMillis();
+
         maxCurrentPagesLoads.acquireUninterruptibly();
+
+        long ioStart = System.currentTimeMillis();
+
+        final List<Record> page;
         try {
             page = dataStorageManager.readPage(tableSpaceUUID, table.uuid, pageId);
-        } catch (DataPageDoesNotExistException ex) {
+        } catch (DataPageDoesNotExistException e) {
             return null;
         } finally {
             maxCurrentPagesLoads.release();
         }
-        long _io = System.currentTimeMillis();
-        DataPage result = buildImmutableDataPage(pageId, page);
-        if (LOGGER.isLoggable(Level.FINEST)) {
-            long _stop = System.currentTimeMillis();
-            LOGGER.log(Level.FINEST, "tmp table " + table.name + ","
-                    + "loaded " + result.size() + " records from page " + pageId
-                    + " in " + (_stop - _start) + " ms"
-                    + ", (" + (_io - _start) + " ms read)");
+
+        long ioStop = System.currentTimeMillis();
+
+        final DataPage result = buildImmutableDataPage(pageId, page);
+
+        if (LOGGER.isLoggable(Level.FINE)) {
+            long stop = System.currentTimeMillis();
+            LOGGER.log(Level.FINE, "table {0}, temporary loaded {1} records from page {2} in {3} ms, ({4} ms read)",
+                    new Object[] { result.size(), pageId, (stop - start), (ioStop - ioStart) });
         }
+
         return result;
     }
 
@@ -1976,12 +1995,12 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
 
         long _start = System.currentTimeMillis();
         long _ioAndLock = 0;
-        AtomicBoolean computed = new AtomicBoolean();
+        BooleanHolder computed = new BooleanHolder(false);
 
         try {
             result = pages.computeIfAbsent(pageId, (id) -> {
                 try {
-                    computed.set(true);
+                    computed.value = true;
                     List<Record> page;
                     maxCurrentPagesLoads.acquireUninterruptibly();
                     try {
@@ -1997,7 +2016,7 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
                     throw new RuntimeException(err);
                 }
             });
-            if (computed.get()) {
+            if (computed.value) {
                 _ioAndLock = System.currentTimeMillis();
 
                 final Page.Metadata unload = pageReplacementPolicy.add(result);
@@ -2018,14 +2037,12 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
             }
             throw new DataStorageManagerException(error);
         }
-        if (computed.get()) {
+        if (computed.value && LOGGER.isLoggable(Level.FINE)) {
             long _stop = System.currentTimeMillis();
             LOGGER.log(Level.FINE,
-                    "table " + table.name + ","
-                    + "loaded " + result.size() + " records from page " + pageId
-                    + " in " + (_stop - _start) + " ms"
-                    + ", (" + (_ioAndLock - _start) + " ms read + plock"
-                    + ", " + (_stop - _ioAndLock) + " ms unlock)");
+                    "table {0}, loaded {1} records from page {2} in {3} ms, ({4} ms read + plock, {5} ms unlock)",
+                    new Object[] { result.size(), pageId, (_stop - _start), (_ioAndLock - _start),
+                            (_stop - _ioAndLock) });
         }
         return result;
     }
@@ -2156,8 +2173,8 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
                 if (lock == null) {
 
                     /*
-                     * Lock the building page if we know that checkpointing page has some dirty record and the building page wasn't
-                     * already locked.
+                     * Lock the building page if we know that checkpointing page has some dirty record and the building
+                     * page wasn't already locked.
                      *
                      * We need to lock building page when the checkpoint page is dirty because we have to push the key
                      * on the PK before pushing it on buildingPage map, with no lock a concurrent reader would find the

